### ISSUE TITLE
1.1.380 reports error when using generics and passing DataclassInstance to another method

### ISSUE THREAD

**Issue by HoriaCY** [microsoft/pyright#9003]
**Describe the bug**
Starting with version 1.1.380, running pyright via CLI on the below code sample reports the following error:

```
14:33 - error: Argument of type "<subclass of type[AbstractBase]* and type[DataclassInstance]>*" cannot be assigned to parameter "t" of type "type[T@decode_dataclass]" in function "decode_dataclass"
    Type "<subclass of type[AbstractBase]* and type[DataclassInstance]>*" is not assignable to type "type[T@decode_dataclass]" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

Downgrading to 1.1.379 reports 0 errors.

This seems like a bug, since before passing  `t` to `decode_dataclass` it specifically checks whether `t` is a dataclass.

Adding the `@dataclass` decorator to `AbstractBase` results in pyright to no longer report this error, however, it being an abstract class, I'd like to avoid this if possible. From our understanding it looks like it should work.

It doesn't look like the changelogs for 1.1.380 mention anything like this specifically, except ```Fixed a bug that results in a false positive error during protocol matching because writable class variables defined in a named tuple or a frozen dataclass were considered read-only.```.

Could the above fix have caused this to be reported as an error inadvertently, or is this intended behaviour?

**Code or Screenshots**

```python
from dataclasses import is_dataclass
from typing import TYPE_CHECKING, Any, cast

if TYPE_CHECKING:
    from dataclasses import DataclassInstance  # type: ignore


class AbstractBase:
    pass


def decode_abstract_base[T: AbstractBase](t: type[T], value: Any) -> T:
    if is_dataclass(t):
        return decode_dataclass(t, value)
    return value


def decode_dataclass[T: DataclassInstance](t: type[T], value: dict[str, Any]) -> T:
    return cast(T, t(**value))
```

**VS Code extension or command-line**
Running pyright via CLI `poetry run pyright` - version 1.1.380 reports error described above, while 1.1.379 reports no errors

**Comment by erictraut** [comment]
Thanks for reporting the bug. This turns out to be the same issue as #9002, which was coincidentally reported just minutes before this bug report was filed. 

I've introduced a fix that will be included in the next release of pyright.

**Comment by HoriaCY** [comment]
Thank you for the prompt response!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381.
